mne/datasets/utils.py

```diff
@@ -426,32 +260,97 @@ def _data_path(path=None, force_update=False, update_path=True, download=True,
-            full_name.append(full)
-        del archive_name
-        # 2. Extract all of the files
-        remove_dir = True
-        for u, fp, an, h, fo in zip(url, folder_path, full_name, hash_,
-                                    folder_orig):
-            _extract(path, name, fp, an, fo, remove_dir)
-            remove_dir = False  # only do on first iteration
-        # 3. Remove all of the archives
-        if remove_archive:
-            for an in full_name:
-                os.remove(op.join(path, an))
-
-        logger.info('Successfully extracted to: %s' % folder_path)
-
-    _do_path_update(path, update_path, key, name)
-    path = folder_path[0]
+
+    # downloader & processors TODO: may want to skip using tqdm during tests?
+    progressbar = True
+    if name == 'fake':
+        progressbar = False
+    downloader = pooch.HTTPDownloader(progressbar=progressbar)  # use tqdm
+    unzip = pooch.Unzip(extract_dir=path)  # to unzip downloaded file
+    untar = pooch.Untar(extract_dir=path)  # to untar downloaded file
+
+    # this processor handles nested tar files
+    nested_untar = pooch.Untar(extract_dir=op.join(path, FOLDER_NAMES[name]))
+
+    # create a map for each dataset name to its corresponding processor
+    # Note: when adding a new dataset, a new line must be added here.
+    processors = dict(
+        bst_auditory=nested_untar,
+        bst_phantom_ctf=nested_untar,
+        bst_phantom_elekta=nested_untar,
+        bst_raw=nested_untar,
+        bst_resting=nested_untar,
+        fake=untar,
+        fieldtrip_cmc=nested_untar,
+        kiloword=untar,
+        misc=untar,
+        mtrf=unzip,
+        multimodal=untar,
+        fnirs_motor=untar,
+        opm=untar,
+        sample=untar,
+        somato=untar,
+        spm=untar,
+        testing=untar,
+        visual_92_categories=untar,
+        phantom_4dbti=unzip,
+        refmeg_noise=unzip,
+        hf_sef_raw=pooch.Untar(
+            extract_dir=path, members=[f'hf_sef/{subdir}' for subdir in
+                                       ('MEG', 'SSS', 'subjects')]),
+        hf_sef_evoked=pooch.Untar(
+            extract_dir=path, members=[f'hf_sef/{subdir}' for subdir in
+                                       ('MEG', 'SSS', 'subjects')]),
+        ssvep=unzip,
+        erp_core=untar,
+        epilepsy_ecog=untar,
+    )
+    # construct the mapping needed by pooch
+    pooch_urls = {ARCHIVE_NAMES[key]: URLS[key] for key in URLS}
+
+    # create the download manager
+    fetcher = pooch.create(
+        path=path,
+        base_url='',    # Full URLs are given in the `urls` dict.
+        version=None,   # Data versioning is decoupled from MNE-Python version.
+        registry=None,  # Registry is loaded from file, below.
+        urls=pooch_urls,
+        retry_if_failed=2  # 2 retries = 3 total attempts
+    )
+    # load the checksum registry
+    registry = files('mne.data').joinpath('dataset_checksums.txt')
```

Comment:
@larsoner are you OK with this use of `importlib_resources.files` here?  I had to add `importlib_resources` to `requirements_testing` for this to work.  (the alternative is `pkg_resources`, which we're supposed to be moving away from)

(FYI, the corresponding `importlib.resources.files()` was added in py3.9 so we can't yet use `importlib` directly for this)

Reply:
> I had to add importlib_resources to requirements_testing for this to work

This is not used only in `tests/*.py` but also in the main `mne` module, so this needs to be in `environment.yml` and `requirements.txt` instead. What's wrong with `op.join(op.dirname(__file__), '..', 'data', 'dataset_checksums.txt')`? Seems clear enough and simple enough to avoid a new dependency...

Reply:
> This is not used only in `tests/*.py` but also in the main `mne` module, so this needs to be in `environment.yml` and `requirements.txt` instead.

that's why I asked (but admittedly I didn't make it clear; what I meant was "for the moment I hacked this in via `requirements_testing` but this will become a new dependency").

> What's wrong with `op.join(op.dirname(__file__), '..', 'data', 'dataset_checksums.txt')`?

1. "The use of `__file__` doesn’t work if your package lives inside a zip file, since in that case this code does not live on the file system" (https://importlib-resources.readthedocs.io/en/latest/using.html#example)  I interpret that to mean that there are potential problems with egg files, although admittedly I haven't dug deep on this.
2. using `__file__` this way is, IMO, hard to read and hard to use (I often struggle with this when I'm trying to interactively run a test line-by-line; I can't just `ctrl+enter` to run a line that defines a path if it has a `dirname(__file__)` in it).
3. `importlib` is part of the standard library and designed for just this kind of use case, and `importlib_resources` is a transitional package with the same API as `importlib.resources` backported to work with older versions of python. in that sense it is future-compatible in that the only thing that will need to change later is to change `_` to `.` in `from importlib_resources import foo`

Reply:
if you want, I could skip the transitional package, try plain `importlib` (will work on py3.9 only) and fallback to `pkg_resources`, similar to what is done here:

https://github.com/mne-tools/mne-python/blob/3aeee826450c823464282657a38c6751d9cdc5a2/doc/conf.py#L898-L906

Reply:
Sure, maybe a new `mne.utils._mne_path()` that returns the path? Then eventually all of our `__file__`s can just use that thing, and that thing can implement the best try/except-based path logic based on our minimum python version